EFCore.CheckConstraints/Internal/EnumCheckConstraintConvention.cs

```diff
@@ -41,34 +42,66 @@ public virtual void ProcessModelFinalizing(IConventionModelBuilder modelBuilder,
-                if (propertyType.IsEnum
-                    && typeMapping != null
-                    && !propertyType.IsDefined(typeof(FlagsAttribute), true)
-                    && property.GetColumnName(tableIdentifier) is { } columnName)
+                if (!propertyType.IsEnum
+                    || typeMapping is null
+                    || propertyType.IsDefined(typeof(FlagsAttribute), true)
+                    || property.GetColumnName(tableIdentifier) is not { } columnName)
                 {
-                    var enumValues = Enum.GetValues(propertyType);
-                    if (enumValues.Length <= 0)
-                    {
-                        continue;
-                    }
+                    continue;
+                }
+
+                var values = Enum.GetValues(propertyType);
+                if (values.Length == 0)
+                {
+                    continue;
+                }
 
-                    sql.Clear();
+                sql.Clear();
 
-                    sql.Append(_sqlGenerationHelper.DelimitIdentifier(columnName));
+                sql.Append(_sqlGenerationHelper.DelimitIdentifier(columnName));
+
+                if (TryParseContiguousRange(values, typeMapping, out var minValue, out var maxValue))
+                {
+                    sql.Append(" BETWEEN ");
+                    sql.Append(minValue);
+                    sql.Append(" AND ");
+                    sql.Append(maxValue);
+                }
+                else
+                {
                     sql.Append(" IN (");
-                    foreach (var item in enumValues)
+                    foreach (var item in values.Cast<object>().Select(typeMapping.GenerateSqlLiteral))
                     {
-                        var value = typeMapping.GenerateSqlLiteral(item);
-                        sql.Append($"{value}, ");
+                        sql.Append(item);
+                        sql.Append(", ");
                     }
 
                     sql.Remove(sql.Length - 2, 2);
-                    sql.Append(")");
-
-                    var constraintName = $"CK_{tableName}_{columnName}_Enum";
-                    entityType.AddCheckConstraint(constraintName, sql.ToString());
+                    sql.Append(')');
                 }
+
+                var constraintName = $"CK_{tableName}_{columnName}_Enum";
+                entityType.AddCheckConstraint(constraintName, sql.ToString());
             }
         }
     }
+
+    private static bool TryParseContiguousRange(IEnumerable values, CoreTypeMapping typeMapping, out decimal minValue, out decimal maxValue)
+    {
+        if (typeMapping.Converter?.ProviderClrType == typeof(string))
+        {
+            minValue = 0;
+            maxValue = 0;
+
+            return false;
+        }
+
+        // we use decimal explicitly here because decimal is a wider number type than all of the valid enum backing types
```

Comment:
I'm not sure sure why the conversion to decimal is needed: can't we just do min/max on the values directly, and then return the SQL literal representations for the min/max from the function (as strings)?

Reply:
No not really; we still need to do `maxValue - minValue` and you can't add or subtract `object`s. We _could_ cast like so: `(dynamic)maxValue - (dynamic)minValue` but there's an exception thrown for the `ulong` backing type. We also can't create an `Expression` because you can't add or subtract `byte` types.

You make an interseting point about using the "SQL literal representations". It is probably valuable to call `typeMapping.GenerateSqlLiteral` even when dealing with number-backing fields.

Reply:
@celluj34 how about making use of the new C# 7 [generic math](https://learn.microsoft.com/en-us/dotnet/standard/generics/math) feature? It was basically built for exactly this kind of thing. It would require us to target the net7.0 TFM, but I think that's fine (EF Core 8.0 will almost certainly target net8.0 anyway).

The one annoying detail is that we're not in a generic context here; so we'd have to use reflection (MakeGenericMethod) to call the generic TryParseContiguousRange method. Shouldn't be a big problem, and would allow you to do whatever math you want inside that function.

Reply:
Using generic math is cool, but I'm running into one problem: `enumValues.Count` is an `int` and no matter how I slice it (without casting or `Convert.ChangeType`) I can't add an `INumber<>` and an `int`.

I've pushed what I have (and all tests are passing) but the concern would be an Enum with more elements than can be represented by an arbitrary number type, the smallest of which being `byte`, with 255 max items.

Reply:
I'm not sure I see the problem... Once the [duplicate enum value problem](https://github.com/efcore/EFCore.CheckConstraints/pull/96#pullrequestreview-1253173613) is addresed, shouldn't a byte-backed enum always have at most count=255?

Reply:
Hmm, you're probably right. Maybe it's not as big of a deal as I think it is. `Convert.ChangeType` is working so I'll keep using it, unless you have a better idea!